### PR TITLE
feat: Install ODF in external mode to Infra cluster

### PR DESCRIPTION
**Pull request by tumido**
Part of: https://github.com/operate-first/operations/issues/475
Depends on: https://github.com/operate-first/apps/pull/2296 https://github.com/operate-first/apps/pull/2360

/hold

Requires credentials to be created in Vault first

/cc @larsks @HumairAK

### CODE REVIEW

File: cluster-scope/overlays/prod/moc/infra/externalsecrets/rook-ceph-external-cluster-details.yaml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+apiVersion: external-secrets.io/v1beta1
+kind: ExternalSecret
+metadata:
+  name: rook-ceph-external-cluster-details
+  namespace: openshift-storage
+spec:
+  secretStoreRef:
+    name: opf-vault-store
+    kind: SecretStore
```
Comment (HumairAK) [inline cluster-scope/overlays/prod/moc/infra/externalsecrets/rook-ceph-external-cluster-details.yaml:9]: this secret store needs to be added for this namespace, and allow vault infra SA to access vault secrets for moc-infra
docs on how to do this: https://www.operate-first.cloud/apps/content/vault_eso/add_external_secrets_to_ns.html#enable-namespace-to-fetch-secrets-from-vault

Comment (tumido) [reply cluster-scope/overlays/prod/moc/infra/externalsecrets/rook-ceph-external-cluster-details.yaml:9]: Resolved, I've totally missed that, thank you!

**Comment by larsks** [comment]
@tumido The credentials get stored in a secret that looks like this:

```
[
  {
    "name": "rook-ceph-mon-endpoints",
    "kind": "ConfigMap",
    "data": {
      "data": "ceph0=<address of ceph mon>",
      "maxMonId": "0",
      "mapping": "{}"
    }
  },
  {
    "name": "monitoring-endpoint",
    "kind": "CephCluster",
    "data": {
      "MonitoringEndpoint": "fake-metrics-exporter",
      "MonitoringPort": "9283"
    }
  },
  {
    "name": "rook-ceph-mon",
    "kind": "Secret",
    "data": {
      "admin-secret": "admin-secret",
      "fsid": "<ceph fsid>",
      "mon-secret": "mon-secret"
    }
  },
  {
    "name": "rook-ceph-operator-creds",
    "kind": "Secret",
    "data": {
      "userID": "client.healthchecker-<cluster_name>-rbd",
      "userKey": "<key for client.healthchecker>"
    }
  },
  {
    "name": "ceph-rbd",
    "kind": "StorageClass",
    "data": {
      "pool": "<cluster_name>_rbd"
    }
  },
  {
    "name": "rook-csi-rbd-node",
    "kind": "Secret",
    "data": {
      "userID": "node-<cluster_name>-rbd",
      "userKey": "<key for node-<cluster_name>-rbd"
    }
  },
  {
    "name": "rook-csi-rbd-provisioner",
    "kind": "Secret",
    "data": {
      "userID": "provisioner-<cluster_name>-rbd",
      "userKey": "<key for provisioner-<cluster_name>-rbd"
    }
  }
]
```

Note this one entry:

```
  {
    "name": "monitoring-endpoint",
    "kind": "CephCluster",
    "data": {
      "MonitoringEndpoint": "fake-metrics-exporter",
      "MonitoringPort": "9283"
    }
  },
```

We found that the operator install would fail if (a)
`monitoring-endpoint` was undefined, and (b) the operator was unable
to contact the prometheus endpoint and extract metrics. We created a
dummy service to provide an appropriate endpoint, since NESE does not
expose (or even run) an appropriate endpoint.

It's not clear that this is necessary once the operator install is
complete.

**Comment by tumido** [comment]
/unhold

Secret is present in vault now.

**Review by larsks — COMMENTED** [review]
This looks fine, but per my comments on slack we may need to ask for some changes on the NESE side. Quoting my entire recent email to @tumido, since the information should probably be here...

---

This is what my original request (to [help@nese.mghpcc.org](mailto:help@nese.mghpcc.org)) for moc-infra storage looked like:

> Hello,
> We would like to request a 1TB volume for the operate first infra cluster ("moc-infra"). This is similar to request #60; we need:
> - A 1TB RBD triple-replicated RDB pool
> - The following auth principals:
>       ceph auth add client.healthchecker-moc-infra \
>               mgr "allow command config" \
>               mon "allow r, allow command quorum_status, allow command version"
>       ceph auth add client.csi-rbd-provisioner-moc-infra \
>               mgr "allow rw" \
>               mon "profile rbd" \
>               osd "profile rbd pool=${poolname}"
>       ceph auth add client.csi-rbd-node-moc-infra \
>               mon "profile rbd" \
>               osd "profile rbd pool=${poolname}"
> 

But it looks like since then the permissions required for ODF have changed; here's a followup I sent recently regarding the NERC infra cluster:

> It looks like ODF under OCP 4.10 needs to be able to run the `ceph mgr dump` command. Based on the permissions that the setup script creates, it looks like we need `caps mgr = "allow command config"` for all client.healthchecker* ceph principals.
> These are the permissions that the setup script creates, which differ somewhat from what we saw with 4.8:
> client.csi-rbd-node
>         key: ...
>         caps: [mon] profile rbd, allow command 'osd blocklist'
>         caps: [osd] profile rbd
> client.csi-rbd-provisioner
>         key: ...
>         caps: [mgr] allow rw
>         caps: [mon] profile rbd, allow command 'osd blocklist'
>         caps: [osd] profile rbd
> client.healthchecker
>         key: ...
>         caps: [mgr] allow command config
>         caps: [mon] allow r, allow command quorum_status, allow command version
>         caps: [osd] allow rwx pool=default.rgw.meta, allow r pool=.rgw.root, allow rw pool=default.rgw.control, allow rx pool=default.rgw.log, allow x pool=default.rgw.buckets.index
> 

And Milan from NESE responded:

> Allowed 'mgr dump' at monitor level. Have some concerns about possible unwanted leaks. Let me know if this works for you so we can proceed with replicating the same config for the healthchecker keys of all three OCP clusters.

So we may need to request some permission changes. 

---

While we have previously configured the NESE storage VLAN on these nodes, it looks like the interface is currently having some problems:

```
$ k get nncp vlan-211-nese
NAME            STATUS
vlan-211-nese   FailedToConfigure
```

That will need to be resolved before the nodes will be able to access the ceph cluster.

I'm going to be out for a few days (back on Tuesday 9/6); @naved001 can probably help out with networking and honestly is probably more careful than I am about some things :smile:.

**Comment by tumido** [comment]
@larsks interesting.. It seems it's not the only network policy failing there. Seems like a more serious networking issue to me...

**Review by SamoKopecky — APPROVED** [review]
/lgtm

**Comment by Gregory-Pereira** [comment]
Please rebase at your earliest convenience.

**Comment by durandom** [comment]
@tumido good work, if it works :) 
I think I get the overall idea, but could you briefly explain where these manifests come from? 
Point to the documentation or another reference used.

**Comment by tumido** [comment]
@durandom I've just followed the docs:

1. Install the operator:
   - https://access.redhat.com/documentation/en-us/red_hat_openshift_data_foundation/4.10/html-single/deploying_openshift_data_foundation_in_external_mode/index#installing-openshift-data-foundation-operator-using-the-operator-hub_ceph-external
   - this gives you `Subscription`, `OperatorGroup`, `Namespace` resources
2. Create the external cluster:
   - https://access.redhat.com/documentation/en-us/red_hat_openshift_data_foundation/4.10/html-single/deploying_openshift_data_foundation_in_external_mode/index#creating-an-openshift-data-foundation-cluster-service-for-external-storage_ceph-external (skipping the script download and apply, since this is handled at NESE)
   - this gives you `StorageCluster` and `Secret`/`ExternalSecret` resources.
   - in order to enable `ExternalSecret` in the namespace you need to follow https://www.operate-first.cloud/apps/content/vault_eso/add_external_secrets_to_ns.html#enable-namespace-to-fetch-secrets-from-vault
3. For convenience when troubleshooting, I've enabled toolbox in `OCSInitialization`:  https://access.redhat.com/documentation/en-us/red_hat_openshift_data_foundation/4.10/html-single/red_hat_openshift_data_foundation_architecture/index#openshift-container-storage-initialization_rhodf

**Pull request merged**